File: gameOfLife/gameOfLife.py
#!/usr/bin/python
# gameOfLife.py (again!)

# Headers
import curses

# Global instance, GL
GL = {}

# Model (State is stored in a global instance GL)
GL['R'] = 0
GL['C'] = 0
GL['r'] = 0
GL['c'] = 0
GL['map'] = [ [' '] * GL['C'] for x in range(GL['R']) ]
GL['running'] = False
def initModel(R,C,r=0,c=0):
    # Set cursor position
    GL['r'] = r
    GL['c'] = c
    # Set dimensions
    GL['R'] = R
    GL['C'] = C
    # Build the map.
    GL['map'] = [ [' '] * GL['C'] for x in range(GL['R']) ]
def moveModelCursor(dr,dc):
    newR = GL['r'] + dr
    newC = GL['c'] + dc
    if newR >= 0 and newR < GL['R'] and newC >= 0 and newC < GL['C']:
        GL['r'] = newR
        GL['c'] = newC
def toggleModel(r,c):
    if GL['map'][r][c] == ' ':
        GL['map'][r][c] = '*'
    elif GL['map'][r][c] == '*':
        GL['map'][r][c] = ' '
    else:
        GL['map'][r][c] = '?'
def toggleModelAtCursor():
    toggleModel(GL['r'],GL['c'])
    
# support methods for tickModel()
def neighborModelCell(r,c,dr,dc):
    R = (r+dr)%GL['R']
    C = (c+dc)%GL['C']
    return (R,C)
def liveModelNeighborCount(r,c):
    ret = 0
    dx = (-1,0,1)
    from itertools import product
    f = lambda r,c,dr,dc : neighborModelCell(r,c,dr,dc)
    dx2 = [ f(r,c,ri,ci) for (ri,ci) in product(dx,dx) if (ri,ci) != (0,0) ]
    return len([ (ri,ci) for (ri,ci) in dx2 if GL['map'][ri][ci] == '*' ])
# Tick model, per Conway's Game of Life.
def tickModel():
    newMap = [ [' '] * GL['C'] for x in range(GL['R']) ]
    for r in range(GL['R']):
        for c in range(GL['C']):
            count = liveModelNeighborCount(r,c)
            if count < 2 or count > 3:
                newMap[r][c] = ' '
            elif count == 3:
                newMap[r][c] = '*'
            elif count == 2:
                newMap[r][c] = GL['map'][r][c]
    GL['map'] = newMap

# View (all ncurses logic remains here)
def initView():
    GL['scr'] = curses.initscr()
    curses.noecho()
    curses.raw()
    GL['scr'].keypad(1)
    
def endView():
    curses.endwin()
    
def getKey():
    return GL['scr'].getch()
    
def drawView():
    GL['scr'].erase()
    R, C = GL['scr'].getmaxyx()
    lowR = (R - GL['R'])/2
    lowC = (C - GL['C'])/2
    
    # Draw the borders
    for r in range(GL['R']):
        GL['scr'].addch(r+lowR,lowC-1,'|')
        GL['scr'].addch(r+lowR,lowC+GL['C'],'|')
    for c in range(GL['C']):
        GL['scr'].addch(lowR-1,lowC+c,'-')
        GL['scr'].addch(lowR+GL['R'],lowC+c,'-')
    
    # Draw the map
    for r in range(GL['R']):
        for c in range(GL['C']):
            GL['scr'].addch(lowR+r,lowC+c,GL['map'][r][c])
    
    # Move the cursor to position as indicated by the model.
    GL['scr'].move(GL['r']+lowR, GL['c']+lowC)
    
    # Finally make sure it is drawn to the screen
    GL['scr'].refresh()
    
# Glue code (Controller?)
def main():
    from sys import argv
    if len(argv) < 3:
        print("Usage: %s [Rows] [Columns]" % argv[0])
        return None
    else:
        initModel(int(argv[1]),int(argv[2]))
        
    try:
        initView()
        GL['running'] = True
        while GL['running']:
            drawView()
            
            # Handle input
            key = getKey()
            # -- exit --
            if key == ord('q') or key == 3:
                GL['running'] = False
            # -- movement --
            elif key == ord('a'):
                moveModelCursor(0,-1)
            elif key == ord('d'):
                moveModelCursor(0, 1)
            elif key == ord('w'):
                moveModelCursor(-1,0)
            elif key == ord('s'):
                moveModelCursor( 1,0)
            # -- toggle cell --
            elif key == ord(' '):
                toggleModelAtCursor()
            # -- tick model --
            elif key == ord('t'):
                tickModel()
            # -- reset model --
            elif key == ord('r'):
                initModel(GL['R'],GL['C'], GL['r'], GL['c'])
            
        endView()
        print("Exited properly.")
    except:
        endView()
        print("Something went wrong.")
    
if __name__ == '__main__':
    main()

File: gameOfLife/gameOfLife.c
#include<stdlib.h>
#include<ncurses.h>
// model.
int running = FALSE;
char** map = NULL;
size_t Row = 0;
size_t Col = 0;
void deleteModelFromCharPtrPtr(char** map);
void deleteModel();
void initModel(size_t r, size_t c);
void randomizeModel(int seed);
int countLiveNeighborsOfModelAt(char** map, int r, int c);
void tickModel();
// Enter UI loop -- all ncurses logic remains here.
void run();
// main
int main(int argc, char* argv[]){
    initModel(10,10);
    randomizeModel(10);
    run();
    return 0;
}

// implementations.
// model.
void deleteModelFromCharPtrPtr(char** map){
    if (map != NULL){
        size_t r, c;
        for (r = 0; r < Row; r++){
            free(map[r]);
        }
        free(map);
    }
}
void deleteModel(){ deleteModelFromCharPtrPtr(map); }
void initModel(size_t r, size_t c){
    int ri, ci;
    
    deleteModel();
    Row = r;
    Col = c;
    
    map = malloc(Row*sizeof(char*));
    for (ri = 0; ri < Row; ri++){
        map[ri] = malloc(Col*sizeof(char));
        for (ci = 0; ci < Col; ci++){
            map[ri][ci] = 0;
        }
    }
}
void randomizeModel(int seed){
    int r, c;
    
    
    srand(seed);
    
    for (r = 0; r < Row; r++){
        for (c = 0; c < Col; c++){
            map[r][c] = rand() % 2;
        }
    }
}
int countLiveNeighborsOfModelAt(char** map, int r, int c){
    int dr, dc;
    int rr, cc;
    
    int ret = 0;
    for (dr = -1; dr <= 1; dr++){
        rr = (r + Row + dr) % Row;
        for (dc = -1; dc <= 1; dc++){
            if ( (dr == dc) && (dc == 0) ) continue;
            
            cc = (c + Col + dc) % Col;
            
            if (map[rr][cc] == 1){
                ret++;
            }
        }
    }
    return ret;
}
void tickModel(){
    int ri, ci;
    
    // create a copy
    char** cpy;
    cpy = malloc(Row*sizeof(char*));
    for (ri = 0; ri < Row; ri++){
        cpy[ri] = malloc(Col*sizeof(char));
        for (ci = 0; ci < Col; ci++){
            cpy[ri][ci] = map[ri][ci];
        }
    }
    
    // Now update the original
    for (ri = 0; ri < Row; ri++){
        for (ci = 0; ci < Col; ci++){
            switch (countLiveNeighborsOfModelAt(cpy,ri,ci)){
                case 2:
                    map[ri][ci] = cpy[ri][ci];
                    break;
                case 3:
                    map[ri][ci] = 1;
                    break;
                default: // less than 2 or more than 3
                    map[ri][ci] = 0;
                    break;
            }
        }
    }
    
    // Now free memory allocated for the copy
    deleteModelFromCharPtrPtr(cpy);
}

// Enter UI loop -- all ncurses logic remains here.
void run(){
    int maxR, maxC;
    int lowR, lowC;
    int r, c;
    char toDraw;
    
    initscr();
    raw();
    noecho();
    keypad(stdscr,TRUE);
    curs_set(0);
    
    running = TRUE;
    while (running){
        // draw.
        erase(); // clear
        
        // calculate borders
        getmaxyx(stdscr,maxR, maxC);
        // leave space for border.
        lowR = (maxR - Row) / 2;
        lowC = (maxC - Col) / 2;
        
        // draw border
        for (r = 0; r < Row; r++){
            mvaddch(lowR+r,lowC-1,'|');
            mvaddch(lowR+r,lowC+Col,'|');
        }
        for (c = 0; c < Col; c++){
            mvaddch(lowR-1,lowC+c,'-');
            mvaddch(lowR+Row,lowC+c,'-');
        }
        mvaddch(lowR-1,lowC+Col,'&');
        mvaddch(lowR-1,lowC-1,'&');
        mvaddch(lowR+Row,lowC+Col,'&');
        mvaddch(lowR+Row,lowC-1,'&');
        
        
        // draw map
        for (r = 0; r < Row; r++){
            for (c = 0; c < Col; c++){
                switch (map[r][c]) {
                    case 0:
                        toDraw = ' ';
                        break;
                    case 1:
                        toDraw = '*';
                        break;
                    default:
                        toDraw = '?';
                        break;
                }
                mvaddch(r+lowR,c+lowC,toDraw);
            }
        }
        refresh();
        
        // handle input.
        switch (getch()){
            case 3: // ctrl-c
                running = FALSE;
                break;
            case ' ':
                tickModel();
                break;
        }
    }
    
    endwin();
}
